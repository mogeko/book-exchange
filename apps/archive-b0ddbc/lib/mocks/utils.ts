import { faker } from "@faker-js/faker";

export function oneOf<T = any>(array: T[]): T {
  return array[randomNum(array.length - 1)];
}

export function someOf<T = any>(array: T[]): T[] {
  const result = array.filter(randomBool);

  return result.length === 0 ? [oneOf(array)] : result;
}

export function languages<T = any>(fn: (lang: string[]) => T): T {
  const langs = [
    "Afrikaans",
    "Arabic",
    "Azerbaijani",
    "Czech",
    "German",
    "Greek",
    "English",
    "Spanish",
    "Farsi",
    "Finnish",
    "French",
    "Georgian",
    "Hebrew",
    "Hrvatski",
    "Hungarian",
    "Armenian",
    "Indonesia",
    "Italian",
    "Japanese",
    "Korean",
    "Latvian",
    "Macedonian",
    "Norwegian",
    "Nepalese",
    "Dutch",
    "Polish",
    "Romanian",
    "Russian",
    "Slovakian",
    "Swedish",
    "Turkish",
    "Ukrainian",
    "Urdu",
    "Vietnamese",
    "Chinese",
  ];

  return fn(langs);
}

export const randomNum = faker.datatype.number;

export const randomBool = faker.datatype.boolean;

export const randomDateRecent = () => {
  return faker.date.recent(10, "2022-01-01T00:00:00.000Z").toISOString();
};

export const randomDatePost = () => {
  return faker.date.past(1000, "2000-01-01T00:00:00.000Z").toISOString();
};
