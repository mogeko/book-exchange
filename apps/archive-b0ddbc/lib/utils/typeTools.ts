type Without<T, U> = { [P in Exclude<keyof T, keyof U>]?: never };

// XOR<{foo: string;}, {bar: number}>
// { foo: "test" } // OK
// { bar: 1 } // OK
// { foo: "test",  bar: 1 } // Error
export type XOR<T, U> = T | U extends object
  ? (Without<T, U> & U) | (Without<U, T> & T)
  : T | U;

// Head<['a', 'b', 'c']> // 'a'
export type Head<T extends any[]> = T extends [infer H, ...infer _] ? H : never;

// Tail<['a', 'b', 'c']> // ['b', 'c']
export type Tail<T extends any[]> = T extends [infer _, ...infer R] ? R : never;

// First<['a', 'b', 'c']> // 'a'
export type First<T extends any[]> = Head<T>;

// Last<['a', 'b', 'c']> // 'c'
export type Last<T extends any[]> = T extends [...infer _, infer L] ? L : never;

// Concat<['a', 'b'], ['c', 'd']> // ['a', 'b', 'c', 'd']
export type Concat<T extends any[], U extends any[]> = [...T, ...U];

// TupleToUnion<['a', 'b']> // 'a' | 'b'
export type TupleToUnion<T extends any[]> = T[number];

// AppendArgument<(a: number, b: string) => number, boolean> // (a: number, b: string, c: boolean) => number
export type AppendArgument<Fn extends (...s: any[]) => any, A> = Fn extends (
  ...s: infer T
) => infer R
  ? (...s: [...T, A]) => R
  : never;

// Flatten<[1, 2, [3, 4], [[[5]]]]> // [1, 2, 3, 4, 5]
export type Flatten<T extends any[]> = T extends [infer First, ...infer Rest]
  ? [...(First extends any[] ? Flatten<First> : [First]), ...Flatten<Rest>]
  : T;
