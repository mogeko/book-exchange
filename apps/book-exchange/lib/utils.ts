import { clsx, type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tap<T>(fn: (x: T) => void) {
  return (x: T) => (fn(x), x);
}

export function range(to: number): number[];
export function range(from: number, to: number): number[];
export function range(from: number, to?: number) {
  return Array.from({ length: to ? to - from : from }, (_, i) => i + from);
}
