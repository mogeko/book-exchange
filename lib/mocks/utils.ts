import { faker } from "@faker-js/faker";

export function oneOf<T = any>(array: T[]): T {
  return array[randomNum(array.length - 1)];
}

export function someOf<T = any>(array: T[]): T[] {
  const result = array.filter(randomBool);

  return result.length === 0 ? [oneOf(array)] : result;
}

export function languages<T = any>(fn: (lang: string[]) => T): T {
  const langs = [
    "Afrikaans",
    "Arabic",
    "Azerbaijani",
    "Czech",
    "German",
    "Greek",
    "English",
    "Spanish",
    "Farsi",
    "Finnish",
    "French",
    "Georgian",
    "Hebrew",
    "Hrvatski",
    "Hungarian",
    "Armenian",
    "Indonesia",
    "Italian",
    "Japanese",
    "Korean",
    "Latvian",
    "Macedonian",
    "Norwegian",
    "Nepalese",
    "Dutch",
    "Polish",
    "Romanian",
    "Russian",
    "Slovakian",
    "Swedish",
    "Turkish",
    "Ukrainian",
    "Urdu",
    "Vietnamese",
    "Chinese",
  ];

  return fn(langs);
}

export const randomNum = faker.datatype.number;

export const randomBool = faker.datatype.boolean;
