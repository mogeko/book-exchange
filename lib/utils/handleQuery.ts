function handleQuery<T extends ParamProps>(url: URL, params: T) {
  if (Object.keys(params).length === 0) return url;
  const queryParams = Object.entries(params)
    .filter(([_, value]) => value)
    .map(([key, value]) => {
      if (Array.isArray(value)) {
        return value
          .filter(Boolean)
          .map((v) => `${key}=${v}`)
          .join("&");
      }

      return `${key}=${value}`;
    })
    .join("&");

  return [url, queryParams].join("?");
}

interface ParamProps {
  [key: string]: any;
}

export type URL = `/api/${string}`;
export type Query = [URL | undefined, ParamProps];

export default handleQuery;
