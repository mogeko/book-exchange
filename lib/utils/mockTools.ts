import { faker } from "@faker-js/faker";

export function oneOf(array: any[]) {
  return array[randomNum(array.length - 1)];
}

export function randomLanguage() {
  return oneOf([
    "Afrikaans",
    "Arabic",
    "Azerbaijani",
    "Czech",
    "German",
    "Greek",
    "English",
    "Spanish",
    "Farsi",
    "Finnish",
    "French",
    "Georgian",
    "Hebrew",
    "Hrvatski",
    "Hungarian",
    "Armenian",
    "Indonesia",
    "Italian",
    "Japanese",
    "Korean",
    "Latvian",
    "Macedonian",
    "Norwegian",
    "Nepalese",
    "Dutch",
    "Polish",
    "Romanian",
    "Russian",
    "Slovakian",
    "Swedish",
    "Turkish",
    "Ukrainian",
    "Urdu",
    "Vietnamese",
    "Chinese",
  ]);
}

export function arrayBy(length: number, value: any) {
  return Array.from({ length: length }, () => value);
}

export const listBy = arrayBy;

export const randomNum = faker.datatype.number;
