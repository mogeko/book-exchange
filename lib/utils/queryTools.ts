function handleQuery<T extends {}>(url: `/${string}`, params: T) {
  if (Object.keys(params).length === 0) return url;
  const queryParams = Object.entries(params)
    .filter(([_, value]) => value)
    .map(([key, value]) => `${key}=${value}`)
    .join("&");
  return [url, queryParams].join("?");
}

export default handleQuery;
