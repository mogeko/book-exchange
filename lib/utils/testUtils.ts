import { render, RenderOptions } from "@testing-library/react";
import { WrapProvider } from "@/pages/_app";

function renderWithProvider(ui: React.ReactElement, options?: RenderOptions) {
  return render(ui, { wrapper: WrapProvider, ...options });
}

export * from "@testing-library/react";
export { renderWithProvider as render };
