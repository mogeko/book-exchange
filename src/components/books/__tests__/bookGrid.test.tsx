import "@testing-library/jest-dom";
import { render, waitFor, screen } from "@/lib/utils/testTools";
import BookGrid from "@/components/books/bookGrid";

describe("bookGrid", () => {
  it("renders a BookGrid", async () => {
    const { container } = render(<BookGrid maxPages={2} limit={3} page={5} />);

    await waitFor(() => {
      expect(screen.getAllByRole("figure")).toHaveLength(3);
    });

    expect(container).toMatchSnapshot();
  });
});
