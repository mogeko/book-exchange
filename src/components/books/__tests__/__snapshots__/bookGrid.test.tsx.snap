// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bookGrid renders a BookGrid 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div
      class="flex flex-wrap gap-6 my-8 w-full"
    >
      <div
        class="relative flex flex-col w-28"
      >
        <figure
          aria-label="Cover Image"
          class="flex justify-center items-center"
        >
          <a
            class="relative w-28 aspect-[2/3]"
            href="/books/bk8992"
          >
            <span
              style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
            >
              <img
                alt="catch"
                data-nimg="fill"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </a>
        </figure>
        <div
          class="flex flex-col text-sm"
        >
          <a
            class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
            href="/books/bk8992"
          >
            <h2
              class="truncate"
            >
              catch
            </h2>
          </a>
          <p
            class="truncate"
          >
            Gracie
          </p>
        </div>
      </div>
      <div
        class="relative flex flex-col w-28"
      >
        <figure
          aria-label="Cover Image"
          class="flex justify-center items-center"
        >
          <a
            class="relative w-28 aspect-[2/3]"
            href="/books/bk8718"
          >
            <span
              style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
            >
              <img
                alt="cheek"
                data-nimg="fill"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </a>
        </figure>
        <div
          class="flex flex-col text-sm"
        >
          <a
            class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
            href="/books/bk8718"
          >
            <h2
              class="truncate"
            >
              cheek
            </h2>
          </a>
          <p
            class="truncate"
          >
            Lonnie
          </p>
        </div>
      </div>
      <div
        class="relative flex flex-col w-28"
      >
        <figure
          aria-label="Cover Image"
          class="flex justify-center items-center"
        >
          <a
            class="relative w-28 aspect-[2/3]"
            href="/books/bk4075"
          >
            <span
              style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
            >
              <img
                alt="rivulet"
                data-nimg="fill"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </a>
        </figure>
        <div
          class="flex flex-col text-sm"
        >
          <a
            class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
            href="/books/bk4075"
          >
            <h2
              class="truncate"
            >
              rivulet
            </h2>
          </a>
          <p
            class="truncate"
          >
            Raegan
          </p>
        </div>
      </div>
    </div>
    <div
      class="btn-group mx-auto"
    >
      <button
        class="btn btn-xs"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 192 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
          />
        </svg>
        <span
          class="sr-only"
        >
          Previous Page
        </span>
      </button>
      <button
        class="btn btn-xs btn-active"
        value="0"
      >
        1
      </button>
      <button
        class="btn btn-xs"
        value="1"
      >
        2
      </button>
      <button
        class="btn btn-xs"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 192 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
          />
        </svg>
        <span
          class="sr-only"
        >
          Next Page
        </span>
      </button>
    </div>
  </div>
</div>
`;
